csr/csr.go

```diff
@@ -99,6 +99,8 @@ type names struct {
 // will be the first SAN that is short enough, which is done only for backwards
 // compatibility with prior Let's Encrypt behaviour. The resulting SANs will
 // always include the original CN, if any.
+//
+// Deprecated: TODO(#7311): Use identifier.FromCSR instead.
```

Comment:
I don't think TODO is quite what we want to do -- this returns a `names` struct which differentiates between the CN and SANs, and puts the SANs in a particular order, so it can't be replaced by a function which returns a simple slice. I think the real TODO is for the `names` struct to either hold identifiers instead of strings, or for it to to have `CN, dnsSANs, ipSANs` as fields.

Reply:
That makes sense; I've modified the TODO accordingly.

I think clarity around handling trusted certs vs. untrusted CSRs is a fairly compelling argument to keep `identifier.FromCert` and `FromCSR` separate. But at the same time, I'm uneasy about those plus `csr.NamesFromCSR` all having very similar logic in three independent implementations. It seems like a tough clarity vs. maintainability tradeoff.

So, to take the opposite approach, we could merge these three into `identifier.FromX509(certOrCsr AnInterface) []ACMEIdentifier, ACMEIdentifier` and put the CN in that second return value, which callers could choose to ignore.